backend/src/halo2_structs.rs

```diff
@@ -0,0 +1,99 @@
+use std::{
+    fs,
+    io::{self, Write},
+    path::Path,
+};
+
+use crate::{Backend, BackendBuilder, Proof};
+use ast::analyzed::Analyzed;
+use number::{DegreeType, FieldElement};
+
+pub struct Halo2Builder;
+impl<T: FieldElement> BackendBuilder<T> for Halo2Builder {
+    fn setup_from_params(&self, size: DegreeType) -> Box<dyn Backend<T>> {
+        Box::new(halo2::Halo2::new(size.try_into().unwrap()))
+    }
+
+    fn read(&self, mut input: &mut dyn io::Read) -> Result<Box<dyn Backend<T>>, io::Error> {
+        Ok(Box::new(halo2::Halo2::read(&mut input)?))
+    }
+}
+#[cfg(feature = "halo2")]
+impl<T: FieldElement> Backend<T> for halo2::Halo2<T> {
+    fn prove(
+        &self,
+        pil: &Analyzed<T>,
+        fixed: &[(&str, Vec<T>)],
+        witness: Option<&[(&str, Vec<T>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        let Some(witness) = witness else {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "witness is missing"));
+        };
+
+        let (proof, fname) = match prev_proof {
+            Some(proof) => (self.prove_aggr(pil, fixed, witness, proof), "proof.bin"),
```

Comment:
I don't think the behavior of the prover should be defined by an Option argument. These two proofs are quite different, so I think they should be different API entries.

Reply:
I understand `prev_proof == None` as the first step of a long chain of proof aggregations, that is why I found it fitting for it to be an argument.